zerver/migrations/0129_realm_emoji_drop_unique_constraint.py

```diff
@@ -0,0 +1,61 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+import os
+import tempfile
+
+from django.conf import settings
+from django.db import migrations, models
+from django.db.backends.postgresql_psycopg2.schema import DatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+from zerver.lib.upload import get_bucket, upload_backend
+
+OLD_ORIG_IMAGE_PATH_TEMPLATE = '{realm_id}/emoji/{emoji_file_name}.original'
+
+def get_image_data(file_path: str) -> bytes:
+    if settings.LOCAL_UPLOADS_DIR is not None:
+        ### Local
+        local_file_path = os.path.join(settings.LOCAL_UPLOADS_DIR, 'avatars', file_path)
+        image_data = open(local_file_path, "rb").read()
+    else:
+        ### S3
+        bucket_name = settings.S3_AVATAR_BUCKET
+        conn = S3Connection(settings.S3_KEY, settings.S3_SECRET_KEY)
+        bucket = get_bucket(conn, bucket_name)
+        key = bucket.get_key(file_path)
+        image_data = force_bytes(key.get_contents_as_string())
+
+    return image_data
+
+def ensure_emoji_images(realm_emoji: RealmEmoji, old_file_name: str) -> None:
+    old_file_path = OLD_ORIG_IMAGE_PATH_TEMPLATE.format(realm_id=realm_emoji.realm.id,
+                                                        emoji_file_name=old_file_name)
+    image_data = get_image_data(old_file_path)
+    with tempfile.TemporaryFile() as emoji_file:
+        emoji_file.write(image_data)
+        emoji_file.seek(0)
+        upload_backend.upload_emoji_image(emoji_file, realm_emoji.filename, realm_emoji.author)
+
+def migrate_realm_emoji_image_files(apps: StateApps, schema_editor: DatabaseSchemaEditor) -> None:
+    RealmEmoji = apps.get_model('zerver', 'RealmEmoji')
+    for realm_emoji in RealmEmoji.objects.all():
+        old_file_name = realm_emoji.filename
+        realm_emoji.filename = realm_emoji.id
+        realm_emoji.save(update_fields=['filename'])
+        ensure_emoji_images(realm_emoji, old_file_name)
```

Comment:
Is the purpose of this migration to rename the files to be named by ID rather than name?

Reply:
Yeah but leaving a copy of older files behind so that we don't break older references.